twenty-first/src/util_types/simple_hasher.rs

```diff
@@ -120,110 +134,112 @@ pub trait Hasher: Sized + Send + Sync + Clone {
 /// In order to hash arbitrary things using a `Hasher`, it must `impl ToDigest<Digest>`
 /// where the concrete `Digest` is what's chosen for the `impl Hasher`. For example, in
 /// order to
-pub trait ToDigest<Digest> {
-    fn to_digest(&self) -> Digest;
+pub trait Hashable<D> {
+    fn to_sequence(&self) -> Vec<D>;
 }
 
-impl ToDigest<Blake3Hash> for PrimeFieldElementFlexible {
-    fn to_digest(&self) -> Blake3Hash {
-        let bytes = bincode::serialize(&self).unwrap();
-        let digest = Blake3Hash(blake3::hash(bytes.as_slice()));
-
-        digest
+impl Hashable<u8> for usize {
+    fn to_sequence(&self) -> Vec<u8> {
+        (0..4)
+            .map(|i| ((*self >> (8 * i)) & 0xff) as u8)
+            .collect::<Vec<u8>>()
     }
 }
 
 // The specification for MMR from mimblewimble specifies that the
 // node count is included in the hash preimage. Representing the
 // node count as a u128 makes this possible
-impl ToDigest<Blake3Hash> for u128 {
-    fn to_digest(&self) -> Blake3Hash {
-        (*self).into()
+
+impl Hashable<u8> for u128 {
+    fn to_sequence(&self) -> Vec<u8> {
+        (0..8)
+            .map(|i| ((*self >> (8 * i)) & 0xff) as u8)
+            .collect::<Vec<u8>>()
     }
 }
 
-impl ToDigest<Vec<BFieldElement>> for u128 {
-    fn to_digest(&self) -> Vec<BFieldElement> {
-        // Only shifting with 63 *should* prevent collissions for all
-        // numbers below u64::MAX
-        vec![
-            BFieldElement::ring_zero(),
-            BFieldElement::ring_zero(),
-            BFieldElement::ring_zero(),
-            BFieldElement::new((self >> 126) as u64),
-            BFieldElement::new(((self >> 63) % BFieldElement::MAX as u128) as u64),
-            BFieldElement::new((self % BFieldElement::MAX as u128) as u64),
-        ]
+impl Hashable<u8> for XFieldElement {
+    fn to_sequence(&self) -> Vec<u8> {
+        let mut array: Vec<u8> = Vec::with_capacity(3 * 8);
+        for i in 0..3 {
+            for j in 0..8 {
+                array.push(((u64::from(self.coefficients[i]) >> (j * 8)) & 0xff) as u8);
+            }
+        }
+        array
```

Comment:
Two elements representing the same value can hash to two different digests with this code. Maybe convert to `canonical_representation` first?

Reply:
`u64::from( BFieldElement ) -> u64`
converts the BFieldElement to canonical representation before casting.